Update python binding for in-place foreach to return `List[Tensor]`

crcrpar opened the pull request:
Fixes #104817

Examples of generated foreach functions in torch/csrc/autograd/generated/python_torch_functionsEverything.cpp

```C++
// _foreach_acos
static PyObject * THPVariable__foreach_acos(PyObject* self_, PyObject* args, PyObject* kwargs)
{
  HANDLE_TH_ERRORS
  static PythonArgParser parser({
    "_foreach_acos(TensorList self)",
  }, /*traceable=*/true);

  ParsedArgs<1> parsed_args;
  auto _r = parser.parse(nullptr, args, kwargs, parsed_args);
  if(_r.has_torch_function()) {
    return handle_torch_function(_r, nullptr, args, kwargs, THPVariableFunctionsModule, "torch");
  }
  // aten::_foreach_acos(Tensor[] self) -> Tensor[]

  auto dispatch__foreach_acos = [](at::TensorList self) -> ::std::vector<at::Tensor> {
    pybind11::gil_scoped_release no_gil;
    return at::_foreach_acos(self);
  };
  return wrap(dispatch__foreach_acos(_r.tensorlist(0)));
  Py_RETURN_NONE;
  END_HANDLE_TH_ERRORS
}

// _foreach_acos_
static PyObject * THPVariable__foreach_acos_(PyObject* self_, PyObject* args, PyObject* kwargs)
{
  HANDLE_TH_ERRORS
  static PythonArgParser parser({
    "_foreach_acos_(TensorList self)",
  }, /*traceable=*/false);

  ParsedArgs<1> parsed_args;
  auto _r = parser.parse(nullptr, args, kwargs, parsed_args);
  if(_r.has_torch_function()) {
    return handle_torch_function(_r, nullptr, args, kwargs, THPVariableFunctionsModule, "torch");
  }
  // aten::_foreach_acos_(Tensor(a!)[] self) -> ()

  auto dispatch__foreach_acos_ = [](at::TensorList self) -> ::std::vector<at::Tensor> {
    pybind11::gil_scoped_release no_gil;
    at::_foreach_acos_(self);
    return self.vec();
  };
  return wrap(dispatch__foreach_acos_(_r.tensorlist(0)));
  Py_RETURN_NONE;
  END_HANDLE_TH_ERRORS
}
```

crcrpar commented:
@pytorchbot rebase

crcrpar commented on aten/src/ATen/native/cuda/ForeachFunctors.cuh:
```diff
@@ -8,6 +8,7 @@ namespace at::native {
 
 namespace {
 
+// TODO(crcrpar): Audit the necessity once in-place foreachs return self.
 // TODO(crcrpar): Handle version bump in codegen.
```
```suggestion
```

crcrpar commented on torchgen/native_function_generation.py:
```diff
@@ -631,11 +631,19 @@ def gen_composite_out_kernel(g: NativeFunctionsGroup) -> Optional[str]:
+        str(g.out.func.name).startswith(prefix)
+        for prefix in ("_foreach", "_fused_adam", "_fused_adamw", "_fused_sgd")
+    ):
+        return_stmt_str = f"return {rets[0]}.vec();"
```
should this be instead?
I think `unpack` for a TensorList returns a `vector<Tensor>` as per https://github.com/pytorch/pytorch/blob/637cf4a3f2cfdd364005681636ca885bdc4d5887/torch/csrc/autograd/VariableTypeManual.cpp#L106-L116
```suggestion
        return_stmt_str = f"return self_;"
```

soulitzer commented on torchgen/api/cpp.py:
```diff
@@ -248,33 +250,43 @@ def returntype_type(t: Type, *, mutable: bool, symint: bool = False) -> CType:
+            assert (
+                not mutable
+            ), "Native functions should never return a mutable tensor list. They should return void."
+        elem = returntype_type(t.elem, mutable=False, is_foreach=is_foreach)
```
I'm not sure we want to special case return_type for foreach. It seems better to either (1) support returning multiple tensor list in general in the dispatcher, or (2) avoiding the dispatcher from having to be aware of this type of schema at all with some kind of special logic at the python level

(2) seems the easier choice. The concern is whether we are okay with having the python schema diverge from aten.  Technically, we already DO have a precedence of python schema diverging from aten schema in factory functions (but that is not super great - there is additional complexity to implement this if we want the cpp API to have parity, and people have been confused about where requires_grad-ness is being handled).

For (1), it appears from @bdhirsh's comment that there is some trickiness to doing this, https://github.com/pytorch/pytorch/pull/76126#discussion_r854570636, and we avoided doing it because method chaining was not a thing for those out= ops anyway.

I'm curious what @ezyang and @bdhirsh think of this. Given that we do care about method chaining in the case of foreach, maybe we should consider supporting functions returning mutable tensor list.

ezyang replied:
I read the issue. @crcrpar, I'd love to understand what you're going for better, because I am not really seeing the benefit for this. As you say in the issue, this isn't enough to actually do method chaining. If you do want the list to be returned in public API, you can add a little wrapper function that re-returns the inputs. So I am puzzled what the value add here is.

crcrpar replied:
One motivation is this comment https://github.com/pytorch/pytorch/pull/104780#discussion_r1256289461. I thought it was okay as I haven't seen strong NO in https://github.com/pytorch/pytorch/issues/104817.

ezyang replied:
Oh, I see, it's not for chaining, it's for renaming the variable as its meaning changes. Hmm, I guess I can't find too much fault in that.

ezyang replied:
Returning to @soulitzer's review, I think I would prefer this to be solely a special case in Python binding generation, that only works for the codegenned bindings, and everything else is kept the same. This would substantially reduce the blast radius of this PR, since you no longer have to update all of the C++ implementations.

ezyang commented:
@albanD as an original participant in the convo at https://github.com/pytorch/pytorch/pull/104780#discussion_r1256289461 I *do* think you should be involved in this review :P

albanD reviewed (commented):
@soulitzer was asking me about this offline but I guess my answer didn't make it here yet? At a high level, my expectation here is that
```
out = torch._foreach_add_(inp1, inp2)
print(out is inp1) # Should be True
```
In particular this is the expectation from the regular Tensor ops.

So I do think that the simplest fix here is at the python or python binding level. Where you can preserve the aten op returning nothing but return the first argument at the python level.

crcrpar commented:
> the simplest fix here is at the python or python binding level.

I'm not sure how I'm supposed to avoid the current diff in aten.
IIUC this would mean I'd make a change into tools/autograd/gen_variable_type so that in-place foreach functions in `VariableType_?.cpp` have return type and return `self.vec()` (or just `self`). One diff would be as follows:

```diff
diff --git a/tools/autograd/gen_variable_type.py b/tools/autograd/gen_variable_type.py
index 4fa0b98c51..5f9773b645 100644
--- a/tools/autograd/gen_variable_type.py
+++ b/tools/autograd/gen_variable_type.py
@@ -958,12 +958,20 @@ def gen_variable_type_func(
             result["type_derived_method_definitions_Default"] = [type_definition]
             result["wrapper_registrations_Default"] = [wrapper_registration]
         else:
+            return_type = cpp.returns_type(f.func.returns, symint=True).cpp_type()
+            if (is_inplace_foreach_with_return := (
+                f.func.kind() == SchemaKind.inplace
+                and name.startswith("_foreach")
+                and is_tensor_list_type(f.func.arguments.self_arg.argument.type)
+            )):
+                return_type = tensorListT
             if not fn.info:
                 key = "Default"
                 type_definition = METHOD_DEFINITION.substitute(
-                    return_type=cpp.returns_type(
-                        f.func.returns, symint=True
-                    ).cpp_type(),
+                    return_type=return_type,
                     type_wrapper_name=type_wrapper_name(f, key),
                     type_definition_body=emit_body(fn, key),
                     formals=formals,
@@ -974,9 +982,7 @@ def gen_variable_type_func(
             else:
                 for key in fn.info.keys():
                     type_definition = METHOD_DEFINITION.substitute(
-                        return_type=cpp.returns_type(
-                            f.func.returns, symint=True
-                        ).cpp_type(),
+                        return_type=return_type,
                         type_wrapper_name=type_wrapper_name(f, key),
                         type_definition_body=emit_body(fn, key),
                         formals=formals,
@@ -2159,4 +2165,6 @@ def emit_body(
         body.append("reset_grad_accumulator(self);")
     if not returns_void:
         body.append(f"return {get_return_value(f)};")
+    if is_inplace_foreach and is_tensor_list_type(f.func.arguments.self_arg.argument.type):
+        body.append(f"return self;")
```

This gives me expected in-place definitions e.g.

```c++
at::TensorList _foreach_acos_(c10::DispatchKeySet ks, at::TensorList self) {
  auto self_ = unpack(self, "self", 0);
...
  return self;
}
```

but it seems that this would violate a rule:

```
terminate called after throwing an instance of 'c10::Error'
  what():
Mismatch in kernel C++ signatures
  operator: aten::_foreach_acos_(Tensor(a!)[] self) -> ()
    registered at /home/mkozuki/ghq/github.com/crcrpar/torch-1/build/aten/src/ATen/RegisterSchema.cpp:6
  kernel 1: void (c10::ArrayRef<at::Tensor>)
    dispatch key: CPU
    registered at /home/mkozuki/ghq/github.com/crcrpar/torch-1/build/aten/src/ATen/RegisterCPU.cpp:31375
  kernel 2: c10::ArrayRef<at::Tensor> (c10::ArrayRef<at::Tensor>)
    dispatch key: Autograd
    registered at /home/mkozuki/ghq/github.com/crcrpar/torch-1/torch/csrc/autograd/generated/VariableType_0.cpp:17451

Exception raised from registerKernel at /home/mkozuki/ghq/github.com/crcrpar/torch-1/aten/src/ATen/core/dispatch/OperatorEntry.cpp:130 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x8e (0x7fde1f7a6b6e in /home/mkozuki/ghq/github.com/crcrpar/torch-1/torch/lib/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0xca (0x7fde1f7a503a in /home/mkozuki/ghq/github.com/crcrpar/torch-1/torch/lib/libc10.so)
frame #2: c10::impl::OperatorEntry::registerKernel(c10::Dispatcher const&, std::optional<c10::DispatchKey>, c10::KernelFunction, std::optional<c10::impl::CppSignature>, std::unique_ptr<c10::FunctionSchema, std::default_delete<c10::FunctionSchema> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x9ae (0x7fddfa0c9a3e in /home/mkozuki/ghq/github.com/crcrpar/torch-1/torch/lib/libtorch_cpu.so)
frame #3: c10::Dispatcher::registerImpl(c10::OperatorName, std::optional<c10::DispatchKey>, c10::KernelFunction, std::optional<c10::impl::CppSignature>, std::unique_ptr<c10::FunctionSchema, std::default_delete<c10::FunctionSchema> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x130 (0x7fddfa0c01b0 in /home/mkozuki/ghq/github.com/crcrpar/torch-1/torch/lib/libtorch_cpu.so)
frame #4: torch::Library::_impl(char const*, torch::CppFunction&&, torch::_RegisterOrVerify) & + 0x328 (0x7fddfa0f9b58 in /home/mkozuki/ghq/github.com/crcrpar/torch-1/torch/lib/libtorch_cpu.so)
frame #5: <unknown function> + 0x366b8fb (0x7fddfc86b8fb in /home/mkozuki/ghq/github.com/crcrpar/torch-1/torch/lib/libtorch_cpu.so)
frame #6: <unknown function> + 0xc2c2a9 (0x7fddf9e2c2a9 in /home/mkozuki/ghq/github.com/crcrpar/torch-1/torch/lib/libtorch_cpu.so)
frame #7: <unknown function> + 0xbb493f (0x7fddf9db493f in /home/mkozuki/ghq/github.com/crcrpar/torch-1/torch/lib/libtorch_cpu.so)
frame #8: <unknown function> + 0x647e (0x7fde211ea47e in /lib64/ld-linux-x86-64.so.2)
frame #9: <unknown function> + 0x6568 (0x7fde211ea568 in /lib64/ld-linux-x86-64.so.2)
frame #10: _dl_catch_exception + 0xe5 (0x7fde20974af5 in /lib/x86_64-linux-gnu/libc.so.6)
frame #11: <unknown function> + 0xdff6 (0x7fde211f1ff6 in /lib64/ld-linux-x86-64.so.2)
frame #12: _dl_catch_exception + 0x88 (0x7fde20974a98 in /lib/x86_64-linux-gnu/libc.so.6)
frame #13: <unknown function> + 0xe34e (0x7fde211f234e in /lib64/ld-linux-x86-64.so.2)
frame #14: <unknown function> + 0x9063c (0x7fde2089063c in /lib/x86_64-linux-gnu/libc.so.6)
frame #15: _dl_catch_exception + 0x88 (0x7fde20974a98 in /lib/x86_64-linux-gnu/libc.so.6)
frame #16: _dl_catch_error + 0x33 (0x7fde20974b63 in /lib/x86_64-linux-gnu/libc.so.6)
frame #17: <unknown function> + 0x9012e (0x7fde2089012e in /lib/x86_64-linux-gnu/libc.so.6)
frame #18: dlopen + 0x48 (0x7fde208906c8 in /lib/x86_64-linux-gnu/libc.so.6)
<omitting python frames>
frame #44: <unknown function> + 0x29d90 (0x7fde20829d90 in /lib/x86_64-linux-gnu/libc.so.6)
frame #45: __libc_start_main + 0x80 (0x7fde20829e40 in /lib/x86_64-linux-gnu/libc.so.6)
frame #46: _start + 0x25 (0x56496402e095 in /home/mkozuki/.pyenv/versions/torchdev1-3.11/bin/python)
```

soulitzer commented:
Modifying autograd kernel generation signature would mean that the dispatcher would need to understand mutable tensor list returns, which we want to avoid. 

Python binding level should refer to something like (See `python_torch_function_*.cpp`):
```cpp
// _foreach_abs_
static PyObject * THPVariable__foreach_abs_(PyObject* self_, PyObject* args, PyObject* kwargs)
{
  HANDLE_TH_ERRORS
  static PythonArgParser parser({
    "_foreach_abs_(TensorList self)",
  }, /*traceable=*/false);

  ParsedArgs<1> parsed_args;
  auto _r = parser.parse(nullptr, args, kwargs, parsed_args);
  if(_r.has_torch_function()) {
    return handle_torch_function(_r, nullptr, args, kwargs, THPVariableFunctionsModule, "torch");
  }
  // aten::_foreach_abs_(Tensor(a!)[] self) -> ()
  
  auto dispatch__foreach_abs_ = [](at::TensorList self) -> void {
    pybind11::gil_scoped_release no_gil;
    at::_foreach_abs_(self);
  };
  dispatch__foreach_abs_(_r.tensorlist(0));
  Py_RETURN_NONE;
  Py_RETURN_NONE;
  END_HANDLE_TH_ERRORS
}
```

I'd start by looking at `tools/autograd/gen_python_functions.py`

crcrpar committed:
return self at python binding level

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
in-place return tensors now

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

ezyang commented:
@crcrpar you should have write access to the repo, due to the recent security incident CI won't run automatically unless your PR is on a branch in pytorch/pytorch repo, so consider rehoming it there so that you don't have to get us to approve and run ci

crcrpar commented:
> @crcrpar you should have write access to the repo, due to the recent security incident CI won't run automatically unless your PR is on a branch in pytorch/pytorch repo, so consider rehoming it there so that you don't have to get us to approve and run ci

@ezyang I'm afraid I don't have that access:

```
~/ghq/github.com/crcrpar/torch-1
 % git remote get-url upstream
git@github.com:pytorch/pytorch.git
~/ghq/github.com/crcrpar/torch-1
 % git push upstream inplace-foreach-with-return
ERROR: Permission to pytorch/pytorch.git denied to crcrpar.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

soulitzer commented on tools/autograd/gen_python_functions.py:
```diff
@@ -1349,7 +1351,28 @@ def go(f: NativeFunction) -> str:
+                str(f.func.name).startswith("_foreach_")
+                and f.func.kind() == SchemaKind.inplace
+                and (
+                    self_arg is not None and is_tensor_list_type(self_arg.argument.type)
```
Would it possible to assert that if we are foreach and inplace that self_arg is None and self is tensorlist?

soulitzer commented on tools/autograd/gen_python_functions.py:
```diff
@@ -1349,7 +1351,28 @@ def go(f: NativeFunction) -> str:
+  {dispatch_callee}({dispatch_args});
+  return self.vec();
+}};
+return wrap(dispatch_{name}({lambda_args}){set_requires_grad});
```
Not sure we actually want to return whatever is returned by dispatch here as this would construct a new tuple. We would just want to return the input as-is in this case as per Alban's comment.

crcrpar committed:
assert self_arg

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
return TensorList

Without `wrap`, I see
```
error: no viable conversion from returned value of type 'at::TensorList' (aka 'ArrayRef<at::Tensor>') to function return type 'PyObject *' (aka '_object *')
      return dispatch__foreach_mul_(_r.tensorlist(0), _r.scalarlist(1));
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

ezyang commented:
@crcrpar you should be added now, sorry about the delay

crcrpar commented:
@ezyang thank you! I'll reopen this pull request after pushing this branch to pytorch/pytorch
